tests/unit/test_charm.py

```diff
@@ -89,16 +94,374 @@ def test_on_config_changed(self):
 
+    def test_start_container(self):
+        """A test for config changed hook."""
+        self.harness.set_leader(True)
+        self.harness.enable_hooks()
+
+        # This should work without an exception.
+        self.start_container()
+
+    def test_on_start(self):
+        """Test on-start event handler."""
+        self.start_container()
+
+        self.harness.charm.on.start.emit()
+
+        self.assertEqual(self.harness.charm.unit.status.name, ActiveStatus.name)
+        self.assertEqual(self.harness.charm.unit.status.message, "")
+
+    def test_on_stop(self):
+        """Test on-stop event handler."""
+        self.start_container()
+
+        self.harness.charm.on.stop.emit()
+
+        self.assertEqual(self.harness.charm.unit.status.name, WaitingStatus.name)
+        self.assertEqual(self.harness.charm.unit.status.message, "service stopped")
+
+    def test_on_update_status(self):
+        """Test on-update-status event handler."""
+        self.start_container()
+
+        self.harness.charm.on.update_status.emit()
+
+        self.assertEqual(self.harness.charm.unit.status.name, ActiveStatus.name)
+        self.assertEqual(self.harness.charm.unit.status.message, "")
+
+    def test_restart_action__success(self):
+        """Test the scenario where `restart` action finished successfully."""
+        self.harness.set_leader(True)
+        self.harness.enable_hooks()
+
+        self.start_container()
+
+        self.harness.run_action("restart")
+
+        self.assertEqual(self.harness.charm.unit.status.name, ActiveStatus.name)
+        self.assertEqual(self.harness.charm.unit.status.message, "")
+
+    def test_schema_upgrade_action__success(self):
+        """Test the scenario where `schema-upgrade` action finishes successfully."""
+        self.harness.set_leader(True)
+        self.harness.enable_hooks()
+
+        self.start_container()
+
+        schema_upgrade_container = self.harness.model.unit.get_container("livepatch-schema-upgrade")
+
+        def container_exists_side_effect(path: str) -> bool:
+            if path == "/usr/local/bin/livepatch-schema-tool":
+                return True
+            return False
+
+        schema_upgrade_container.exists = Mock(side_effect=container_exists_side_effect)
+
+        def container_exec_side_effect(command: List[str]):
+            self.assertEqual(
+                command,
+                [
+                    "/usr/local/bin/livepatch-schema-tool",
+                    "upgrade",
+                    "/etc/livepatch/schema-upgrades",
+                    "--db",
+                    "postgresql://123",
+                ],
+            )
+            process_mock = Mock()
+            process_mock.wait_output.side_effect = lambda: (None, None)
+            return process_mock
+
+        schema_upgrade_container.exec = Mock(side_effect=container_exec_side_effect)
+
+        self.harness.run_action("schema-upgrade")
```

Comment:
Only container's `exists` and `exec` methods are replace with a mock. Since we have the objects we can just replace the methods with mocks, instead of patching. I'm not sure if answered your question, though.

Reply:
Okay, I'll need to read it again to understand, was just curious how the mock was being done